typed-racket-lib/typed-racket/rep/rep-utils.rkt

```diff
@@ -24,159 +25,245 @@
+  ;; Rep-values : Rep -> (listof any/c)
+  (Rep-values Rep)
+  ;; is there a simple, structural description for the variances
+  ;; of this Rep's fields? (currently only used for 'structural' types,
+  ;; see type-rep.rkt
+  ;; Rep-constructor : Rep -> (any ... -> Rep)
+  (Rep-variances Rep)
+  ;; The intended constructor for this Rep.
+  ;; i.e. (equal? ((Rep-constructor rep) (Rep-values rep)) rep) = #t
+  ;; Rep-constructor : Rep -> (any ... -> Rep)
+  (Rep-constructor Rep)
+  ;; can this Rep contain free type variables?
+  ;; (i.e. 'F' types from rep/type-rep.rkt)
+  ;; Rep-free-ty-vars-fun : Rep -> free-vars
+  (Rep-free-vars Rep)
+  ;; can this Rep contain free dotted type variables (idxs)?
+  ;; (e.g. things like ListDots, etc rep/type-rep.rkt
+  ;;  which have an arity/structure which depends on instantiation)
+  ;; Rep-free-ty-idxs-fun : Rep -> free-dotted-vars
+  (Rep-free-idxs Rep)
+  ;; is this Rep mappable?
+  ;; (i.e. can we traverse it w/ applying a function to
+  ;;  the fields? a lá map for lists)
+  ;; Rep-fmap : Rep procedure -> Rep
+  (Rep-fmap Rep f)
+  ;; is this Rep walkable?
+  ;; (i.e. can we traverse it w/ some effectful function
+  ;;  a lá for-each for lists)
+  ;; Rep-walk-fun : Rep procedure -> void
+  (Rep-for-each Rep f))
+
+;; used internally when generating gen:Rep method definitions
+;; so that we don't have to mess around w/ 'define/generic'
+(define-syntax free-vars* (make-rename-transformer #'Rep-free-vars))
+(define-syntax free-idxs* (make-rename-transformer #'Rep-free-idxs))
+
+;; A variant-unique fixnum.
+;; Rep-uid : Rep -> fixnum
+(define-values (prop:uid Rep-uid)
+  (let-values ([(prop _ acc) (make-struct-type-property 'uid)])
+    (values prop acc)))
+
+
+(define-values (prop:mask raw-mask)
+  (let-values ([(prop _ acc) (make-struct-type-property 'mask)])
+    (values prop acc)))
+
+;; Type-mask : Rep -> fixnum
+(define-syntax-rule (mask rep)
+  (let ([mask (raw-mask rep)])
+    (if (procedure? mask)
+        (mask rep)
+        mask)))
 
 
 ;;************************************************************
 ;; Rep Declaration Syntax Classes
 ;;************************************************************
-(define (make-counter!)
-  (let ([state 0])
-    (λ () (begin0 state (set! state (unsafe-fx+ 1 state))))))
-
-(define count! (make-counter!))
-(define id-count! (make-counter!))
-
-(define identifier-table (make-free-id-table))
-
-(define (hash-id id)
-  (free-id-table-ref!
-   identifier-table
-   id
-   (λ () (let ([c (id-count!)])
-           (free-id-table-set! identifier-table id c)
-           c))))
+(define-values (next-uid! get-uid-count)
+  (let ([state 0]
+        [finalized? #f])
+    (values (λ () (if finalized?
+                      (int-err "next-uid! called after uid count finalized!")
+                      (begin0 state (set! state (add1 state)))))
+            (λ () (set! finalized? #t) state))))
+
+(define-syntax (define-switch stx)
```

Comment:
Seems like a lot of this could be a separate file (with a little documentation).